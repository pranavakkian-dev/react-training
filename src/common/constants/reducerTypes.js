/* Common reducer */
export const LOGIN = 'LOGIN';
export const RESET = 'RESET';

/* Faculty reducer */
export const ADD_FACULTY = 'ADD_FACULTY';
export const FACULTY_RESET = 'FACULTY_RESET';
export const DELETE_FACULTY = 'DELETE_FACULTY';
