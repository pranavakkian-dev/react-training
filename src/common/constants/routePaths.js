export const ROOT_PATH = '/';
export const FACULTY_PATH = '/faculty';
