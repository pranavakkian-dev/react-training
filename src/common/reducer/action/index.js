import {
  LOGIN, RESET,
} from '../../constants/reducerTypes';

export const handleLogin = (username) => ({
  type: LOGIN,
  payload: username,
});

export const handleReset = () => ({
  type: RESET,
});
