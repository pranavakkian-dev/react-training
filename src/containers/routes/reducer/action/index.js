import {
  LOGIN, RESET,
} from '../../../../common/constants/reducerTypes';

export const handleLogin = (username) => ({
  type: LOGIN,
  payload: username,
});

export const handleReset = () => ({
  type: RESET,
});
