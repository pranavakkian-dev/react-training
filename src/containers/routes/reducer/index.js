import initialState from './initialState';
import {
  LOGIN, RESET,
} from '../../../common/constants/reducerTypes';

const reducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case LOGIN:
      return { ...state, loggedIn: true, userName: payload };
    case RESET:
      return initialState;
    default:
      return state;
  }
};

export default reducer;
