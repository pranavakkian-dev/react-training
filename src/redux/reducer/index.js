import initialState from './initialState';
import {
  LOGIN, RESET,
} from '../../common/constants/reducerTypes';

const Reducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case LOGIN:
      return { ...state, loggedIn: true, userName: payload };
    case RESET:
      return initialState;
    default:
      return state;
  }
};

export default Reducer;
