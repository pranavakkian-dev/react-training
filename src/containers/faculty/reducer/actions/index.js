import { FACULTY_RESET, ADD_FACULTY, DELETE_FACULTY } from '../../../../common/constants/reducerTypes';

export const handleAddFaculty = (faculty) => ({
  type: ADD_FACULTY,
  payload: faculty,
});

export const handleDeleteFaculty = (faculty) => ({
  type: DELETE_FACULTY,
  payload: faculty,
});

export const handleFacultyReset = () => ({
  type: FACULTY_RESET,
});
