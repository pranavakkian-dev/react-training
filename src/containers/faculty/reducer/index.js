import initialState from './initialState';
import { FACULTY_RESET, ADD_FACULTY, DELETE_FACULTY } from '../../../common/constants/reducerTypes';

const reducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case ADD_FACULTY: {
      const { faculties } = state;
      return { ...state, faculties: [...faculties, { ...payload }] };
    }
    case DELETE_FACULTY: {
      const { faculties } = state;
      const updatedFaculties = faculties.filter((faculty) => faculty !== payload);
      return { ...state, faculties: [...updatedFaculties] };
    }
    case FACULTY_RESET:
      return initialState;
    default:
      return state;
  }
};

export default reducer;
