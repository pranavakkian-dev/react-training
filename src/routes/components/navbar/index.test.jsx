import { render, screen } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import userEvent from '@testing-library/user-event';
import NavBar from './index';
import { FACULTY_PATH, ROOT_PATH } from '../../../common/constants/routePaths';
import { handleReset } from '../../../common/reducer/action';

describe('NavBar', () => {
  const history = createMemoryHistory();
  const mockedDispatch = jest.fn();

  test('Component rendered successfully!', () => {
    const { container } = render(
      <Router history={history}>
        <NavBar dispatch={mockedDispatch} />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Faculty page navigation successfully!', () => {
    render(
      <Router history={history}>
        <NavBar dispatch={mockedDispatch} />
      </Router>,
    );
    expect(history.location.pathname).toBe(ROOT_PATH);
    userEvent.click(screen.getByRole('link', { name: 'Faculty' }));
    expect(history.location.pathname).toBe(FACULTY_PATH);
  });

  test('Home page navigation successfully!', () => {
    render(
      <Router history={history}>
        <NavBar dispatch={mockedDispatch} />
      </Router>,
    );
    expect(history.location.pathname).toBe(FACULTY_PATH);
    userEvent.click(screen.getByRole('link', { name: 'Home' }));
    expect(history.location.pathname).toBe(ROOT_PATH);
  });

  test('Logout button working properly!', () => {
    render(
      <Router history={history}>
        <NavBar dispatch={mockedDispatch} />
      </Router>,
    );
    userEvent.click(screen.getByRole('button', { name: 'Logout' }));
    expect(mockedDispatch).toHaveBeenNthCalledWith(1, handleReset());
  });
});
