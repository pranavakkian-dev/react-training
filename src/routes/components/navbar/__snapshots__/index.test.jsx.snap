// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Component rendered successfully! 1`] = `
<div>
  <nav>
    <ul>
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/faculty"
        >
          Faculty
        </a>
      </li>
      <li>
        <button
          type="button"
        >
          Logout
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
